decoder.go

```diff
@@ -90,10 +95,10 @@ func (d *Decoder) DecodeTag() (tag int, wireType WireType, err error) {
 	}
 	v, n, err := DecodeVarint(d.p[d.offset:])
 	if err != nil {
-		return 0, -1, err
+		return 0, -1, fmt.Errorf("invalid data at byte %d: %w", d.offset, err)
 	}
-	if n < 1 || n > MaxTagValue {
-		return 0, -1, ErrInvalidFieldTag
+	if n < 1 || v < 1 || v > MaxTagValue {
```

Comment:
Note: this line is the actual bug fix.

The issue was that if the encoded data at the current offset is not a valid field encoding (in the specific example that exposed this is was 6 bytes of 0x0), we were not verifying that the value for `tag` was valid.  We were instead comparing the number of bytes read (`n`) against `MaxTagValue`.

Because we were not returning an error, the consuming code would continue reading and eventually copy the "bad" data into the `unknownFields` field on the Go struct.  Later, calling `.String()` on the message would panic when the `prototext` package tries to treat those invalid bytes as correctly encoded Protobuf field data.